### ISSUE TITLE
Disposing socket on Linux does not interrupt poll/select

### ISSUE THREAD

**Issue by drieseng** [dotnet/corefx#31368]
On Linux, disposing a socket on a given thread does not interrupt poll/select on another thread.

To reproduce, compile & run the following code and press enter when prompted:

```cs
// Comment this define to test using Socket.Select(...)
#define POLL

using System;
using System.Net;
using System.Net.Sockets;
using System.Threading.Tasks;

class Program
{
    static void Main()
    {
        var endPoint = new IPEndPoint(IPAddress.Loopback, 8080);
        var listener = new TcpListener(endPoint);
        listener.Start();

        var client = new Socket(endPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp) {NoDelay = true};
        client.Connect(endPoint);

        var listenerTask = Task.Factory.StartNew(s =>
            {
                var socket = (Socket) s;

#if POLL
                if (socket.Poll(-1, SelectMode.SelectRead))
                {
                    if (!socket.Connected)
                    {
                        Console.WriteLine("Connection closed.");
                    }
                }
#else
                var readSockets = new System.Collections.Generic.List<Socket> {socket};
                Socket.Select(readSockets, null, null, -1);
                if (!socket.Connected)
                {
                    Console.WriteLine("Connection closed.");
                }
#endif
            }, client, TaskCreationOptions.LongRunning);

        Console.WriteLine("Press enter to dispose the socket.");
        Console.ReadLine();

        Console.WriteLine("Disposing socket...");
        client.Dispose();
        Console.WriteLine("Socket disposed.");

        if (!listenerTask.Wait(5000))
            Console.WriteLine("Socket did not break out of Select call within 5 seconds.");

        Console.WriteLine("Done.");
    }
}
```

### Expected result:
The following output is written to stdout:

> Disposing socket...
> Connection closed.
> Socket disposed.
> Done.

This is the result you get on the desktop CLR, and .NET Core on Windows.

### Actual result:

> Disposing socket...
(hang)

### Info

.NET Core version: _2.1.302_
OS: _Ubuntu 14.04_

**Comment by wfurt** [comment]
#26898 should resolve this and you can call shutdown() as workaround. As @stephentoub mentioned, we are trying to bend OS behavior where blocking IO operation behaves different on Linux/MacOS.

**Comment by wfurt** [comment]
I verified that https://github.com/dotnet/corefx/pull/32087 fixes this issue.
In mean time you can workaround it by adding Shutdown() before the Dispose() 

```c#
        Console.WriteLine("Disposing socket...");
+       client.Shutdown(SocketShutdown.Both);
        client.Dispose();
        Console.WriteLine("Socket disposed.");

```

**Comment by wfurt** [comment]
fixed by #38804
